src/types/CodepointWidthDetector.cpp

```diff
@@ -9,392 +9,358 @@ namespace
-CodepointWidthDetector::CodepointWidthDetector() noexcept :
-    _fallbackCache{},
-    _pfnFallbackMethod{}
-{
-}
-
 // Routine Description:
 // - returns the width type of codepoint as fast as we can by using quick lookup table and fallback cache.
 // Arguments:
 // - glyph - the utf16 encoded codepoint to search for
 // Return Value:
 // - the width type of the codepoint
-CodepointWidth CodepointWidthDetector::GetWidth(const std::wstring_view glyph) const
+CodepointWidth CodepointWidthDetector::GetWidth(const std::wstring_view& glyph) noexcept
 {
-    THROW_HR_IF(E_INVALIDARG, glyph.empty());
-    if (glyph.size() == 1)
-    {
-        // We first attempt to look at our custom quick lookup table of char width preferences.
-        const auto width = GetQuickCharWidth(glyph.front());
+    char32_t codepoint = 0;
 
-        // If it's invalid, the quick width had no opinion, so go to the lookup table.
-        if (width == CodepointWidth::Invalid)
-        {
-            return _lookupGlyphWidthWithCache(glyph);
-        }
-        // If it's ambiguous, the quick width wanted us to ask the font directly, try that if we can.
-        // If not, go to the lookup table.
-        else if (width == CodepointWidth::Ambiguous)
-        {
-            if (_pfnFallbackMethod)
-            {
-                return _checkFallbackViaCache(glyph) ? CodepointWidth::Wide : CodepointWidth::Ambiguous;
-            }
-            else
-            {
-                return _lookupGlyphWidthWithCache(glyph);
-            }
-        }
-        // Otherwise, return Width as it is.
-        else
-        {
-            return width;
-        }
-    }
-    else
+    switch (glyph.size())
     {
-        return _lookupGlyphWidthWithCache(glyph);
+    case 1:
+        codepoint = til::at(glyph, 0);
+        break;
+    case 2:
+        codepoint = (til::at(glyph, 0) & 0x3FF) << 10;
+        codepoint |= til::at(glyph, 1) & 0x3FF;
+        codepoint += 0x10000;
+        break;
+    default:
+        codepoint = 0;
+        break;
     }
-}
 
-// Routine Description:
-// - checks if wch is wide. will attempt to fallback as much possible until an answer is determined
-// Arguments:
-// - wch - the wchar to check width of
-// Return Value:
-// - true if wch is wide
-bool CodepointWidthDetector::IsWide(const wchar_t wch) const noexcept
-{
-    try
+    if (codepoint < 0x80)
     {
-        return IsWide({ &wch, 1 });
+        return CodepointWidth::Narrow;
     }
-    CATCH_LOG();
 
-    return true;
+    // The return value of _lookupGlyphWidth coincides with the enum value of CodepointWidth
+    // on purpose to allow for this easy conversion to happen. Optimally, we should probably
+    // remove CodepointWidth altogether to allow for ZWJs and other characters.
+    static_assert(WI_EnumValue(CodepointWidth::Narrow) == 1);
+    static_assert(WI_EnumValue(CodepointWidth::Wide) == 2);
+    return static_cast<CodepointWidth>(_lookupGlyphWidth(codepoint, glyph));

```

Comment:
This code all the way to the end of the file is a complete rewrite. You can try reading it in the diff viewer, but there's unfortunately nothing to compare it with in the old code, outside of a few simple lines/comments. 😓